### PR TITLE
[tune] Ray Tune API cleanup

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Remove rllib dep: trainable is now a standalone abstract class that can be easily subclassed.

Clean up hyperband: fix debug string and add an example.

Remove YAML api / ScriptRunner: this was never really used.

Move ray.init() out of run_experiments(): This provides greater flexibility and should be less confusing since there isn't an implicit init() done there. Note that this is a breaking API change for tune.

## Related issue number

https://github.com/ray-project/ray/issues/1429

### CODE REVIEW

File: doc/source/tune.rst [hunk 1]
```diff
@@ -30,6 +31,7 @@ Getting Started
 
     register_trainable("my_func", my_func)
 
+    ray.init()
     run_experiments({
```
Comment (richardliaw) [inline doc/source/tune.rst]: I believe you need to modify `rllib.rst` too

**Review by richardliaw — COMMENTED** [review]
There are quite a few mentions of `run_experiments` in the CARLA examples...

File: doc/source/tune.rst (hunk 1 above)
Comment (ericl) [reply doc/source/tune.rst]: Fixed

**Review by richardliaw — COMMENTED** [review]
One last one... `examples/custom_env/custom_env.py`

**Comment by ericl** [comment]
Sorry for merging the other PR here, thought that was best since Travis isn't very happy right now.

File: python/ray/tune/hyperband.py [hunk 2]
```diff
@@ -295,7 +297,9 @@ def current_trials(self):
 
     def continue_trial(self, trial):
         result = self._live_trials[trial]
-        if self._get_result_time(result) < self._cumul_r:
+        if self._get_result_time(result) > self._max_t:
```
Comment (ericl) [inline python/ray/tune/hyperband.py]: @richardliaw is this the right way to limit to max t? If not we can leave it for another pr.

File: python/ray/tune/trainable.py [hunk 3]
```diff
@@ -22,7 +95,47 @@ def train(self):
...
+            neg_mean_loss=neg_loss,
+            pid=os.getpid(),
+            hostname=os.uname()[1])
+
+        self._result_logger.on_result(result)
+
+        return result
 
     def save(self):
         """Saves the current model state to a checkpoint.
```
Comment (richardliaw) [inline python/ray/tune/trainable.py]: also same comment as `train`

File: python/ray/tune/hyperband.py [hunk 4]
```diff
@@ -372,7 +376,9 @@ def __repr__(self):
         status = ", ".join([
             "n={}".format(self._n),
             "r={}".format(self._r),
-            "progress={}".format(self.completion_percentage())
+            "completed={}%".format(int(100 * self.completion_percentage()))
             ])
-        trials = ", ".join([t.status for t in self._live_trials])
-        return "Bracket({})[{}]".format(status, trials)
+        counts = collections.defaultdict(int)
```
Comment (richardliaw) [inline python/ray/tune/hyperband.py]: there is `Counter` from collections that does the same thing, by the way

File: python/ray/tune/hyperband.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/tune/hyperband.py]: it's a little bit more involved than this..

File: python/ray/tune/__init__.py [hunk 5]
```diff
@@ -6,13 +6,10 @@
 from ray.tune.tune import run_experiments
 from ray.tune.registry import register_env, register_trainable
 from ray.tune.result import TrainingResult
-from ray.tune.script_runner import ScriptRunner
 from ray.tune.trainable import Trainable
 from ray.tune.variant_generator import grid_search
 
 
-register_trainable("script", ScriptRunner)
```
Comment (richardliaw) [inline python/ray/tune/__init__.py]: why does this no longer need to be here?

**Review by richardliaw — CHANGES_REQUESTED** [review]
Some doc comments

File: doc/source/tune.rst [hunk 6]
```diff
@@ -85,7 +87,7 @@ To use rllab's VisKit (you may have to install some dependencies), run:
 
 Finally, to view the results with a `parallel coordinates visualization <https://en.wikipedia.org/wiki/Parallel_coordinates>`__, open `ParalleCoordinatesVisualization.ipynb <https://github.com/ray-project/ray/blob/master/python/ray/tune/ParallelCoordinatesVisualization.ipynb>`__ as follows and run its cells:
 
-::
+.. code-block:: bash
```
Comment (richardliaw) [inline doc/source/tune.rst]: Cathy did suggest offering a visualization here

File: python/ray/tune/function_runner.py [hunk 7]
```diff
@@ -85,97 +76,48 @@ def run(self):
...
-        raise TuneError("Unable to import file at {}".format(file_path))
-    return getattr(external_file, function_name)
-
-
-class ScriptRunner(Agent):
-    """Agent that runs a user script returning training results."""
-
-    _agent_name = "script"
+class FunctionRunner(Trainable):
+    """Trainable that runs a user function returning training results.
```
Comment (richardliaw) [inline python/ray/tune/function_runner.py]: Would be nice to document the following:
1.  the fact that functions are run on a separate thread here
2. some form of example usage
3. why save/restore are hard to implement

File: doc/source/tune.rst (hunk 6 above)
Comment (ericl) [reply doc/source/tune.rst]: Let's do that as a separate task.

File: python/ray/tune/function_runner.py (hunk 7 above)
Comment (ericl) [reply python/ray/tune/function_runner.py]: This is an internal class, so this is not necessary.

File: python/ray/tune/hyperband.py (hunk 2 above)
Comment (ericl) [reply python/ray/tune/hyperband.py]: Ok, let me revert this then.

File: python/ray/tune/hyperband.py (hunk 4 above)
Comment (ericl) [reply python/ray/tune/hyperband.py]: Changed

File: python/ray/tune/trainable.py [hunk 8]
```diff
@@ -2,18 +2,91 @@
...
+from ray.tune import TuneError
+from ray.tune.logger import UnifiedLogger
+from ray.tune.result import DEFAULT_RESULTS_DIR
+
 
 class Trainable(object):
-    """Interface for trainable models, functions, etc.
+    """Abstract class for trainable models, functions, etc.
+
+    A call to ``train()`` on a trainable will execute one logical iteration of
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/trainable.py]: it shows up nicer on readthedocs.

File: python/ray/tune/trainable.py [hunk 9]
```diff
@@ -2,18 +2,91 @@
...
+        logdir (str): Directory in which training outputs should be placed.
+        registry (obj): Tune object registry which holds user-registered
+            classes and objects by name.
+    """
+
+    def __init__(self, config={}, registry=None, logger_creator=None):
+        """Initialize an Trainable.
+
+        Subclasses should prefer defining _setup() instead of overriding
+        __init__() directly.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/trainable.py]: Updated

File: python/ray/tune/trainable.py [hunk 10]
```diff
@@ -2,18 +2,91 @@
...
+        self._initialize_ok = False
+        self._experiment_id = uuid.uuid4().hex
+        self.config = config
+        self.registry = registry
+
+        if logger_creator:
+            self._result_logger = logger_creator(self.config)
+            self.logdir = self._result_logger.logdir
+        else:
+            logdir_suffix = datetime.today().strftime("%Y-%m-%d_%H-%M-%S")
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/trainable.py]: Changed to prefix.

File: python/ray/tune/trainable.py [hunk 11]
```diff
@@ -2,18 +2,91 @@
...
+            self._result_logger = UnifiedLogger(self.config, self.logdir, None)
+
+        self._iteration = 0
+        self._time_total = 0.0
+        self._timesteps_total = 0
+        self._setup()
+        self._initialize_ok = True
 
     def train(self):
         """Runs one logical iteration of training.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/trainable.py]: Done

File: python/ray/tune/trainable.py (hunk 3 above)
Comment (ericl) [reply python/ray/tune/trainable.py]: Done

File: test/trial_scheduler_test.py [hunk 12]
```diff
@@ -493,4 +494,5 @@ def testJumpingTime(self):
 
 
 if __name__ == "__main__":
+    rllib._register_all()
```
[the remark this answers is not in the document]
Comment (ericl) [reply test/trial_scheduler_test.py]: Let me move this into the Trial file instead. Basically we still need to import rllib in order to register the agents

**Review by ericl — COMMENTED** [review]
Updated

File: python/ray/tune/tune.py [hunk 13]
```diff
@@ -30,6 +30,9 @@ def _make_scheduler(args):
 
 
 def run_experiments(experiments, scheduler=None):
+    # Make sure rllib agents are registered
+    from ray import rllib  # noqa # pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/tune.py]: Importing rllib actually does not have any hard dependencies -- if some dependency is not installed you just get a warning on the console.

[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/tune.py]: OK, I see. I'm trying to import `run_experiments` on a clean env.

I'm getting an import error telling me that I still need to install TensorFlow (for the Logger) - this should be documented somewhere. Could also just disable the TensorBoard logger if Tensorflow not installed. This can be done on a separate PR (or in the current one)

Also, if a user is trying to use `tune` out of the box (say for HyperBand), it probably makes sense to preface the warnings with something like `Trying to import default rllib agents...` instead of a sudden block of warnings.

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/tune.py]: That's fine for now, we can improve this in a separate PR.

**Review by richardliaw — APPROVED** [review]
overall lgtm

**Comment by richardliaw** [comment]
`trial_runner_test.py` is erroring

**Comment by ericl** [comment]
Builds look good.
